src/installer/managed/Microsoft.NET.HostModel/MachO/CodeSignature.cs

```diff
@@ -0,0 +1,155 @@
+
+using System;
+using System.IO;
+using System.IO.MemoryMappedFiles;
+using System.Linq;
+using System.Runtime.InteropServices;
+
+namespace Microsoft.NET.HostModel.MachO;
+
+internal class CodeSignature
+{
+    private (EmbeddedSignatureHeader Header, long FileOffset) _embeddedSignature;
+    private (CodeDirectoryHeader Header, long FileOffset) _codeDirectory;
+    private (byte[] Identifier, long FileOffset) _identifierPtr;
+    private (byte[] Blob, long FileOffset) _cdHashes;
+    private (RequirementsBlob Header, long FileOffset) _requirementsBlob;
+    private (CmsWrapperBlob Header, long FileOffset) _cmsWrapperBlob;
+    private bool _unrecognizedFormat;
+
+    public uint Size => _embeddedSignature.Header.Size;
+
+    private CodeSignature() { }
+
+    public static bool AreEquivalent(CodeSignature a, CodeSignature b)
+    {
+        if (a is null ^ b is null)
+            return false;
+        if (a is null && b is null)
+            return true;
+        if (a._unrecognizedFormat || b._unrecognizedFormat)
+            return false;
+        if (!a._embeddedSignature.Equals(b._embeddedSignature))
+            return false;
+        if (!a._codeDirectory.Equals(b._codeDirectory))
+            return false;
+        if (!a._identifierPtr.Identifier.SequenceEqual(b._identifierPtr.Identifier))
+            return false;
+
+        var aSpecialSlotHashes = a._cdHashes.Blob.AsSpan(0, (int)MachObjectFile.SpecialSlotCount * MachObjectFile.DefaultHashSize);
+        var bSpecialSlotHashes = b._cdHashes.Blob.AsSpan(0, (int)MachObjectFile.SpecialSlotCount * MachObjectFile.DefaultHashSize);
+        if (!aSpecialSlotHashes.SequenceEqual(bSpecialSlotHashes))
+            return false;
+        var aCodeHashes = a._cdHashes.Blob.AsSpan(((int)MachObjectFile.SpecialSlotCount + 1) * MachObjectFile.DefaultHashSize);
+        var bCodeHashes = b._cdHashes.Blob.AsSpan(((int)MachObjectFile.SpecialSlotCount + 1) * MachObjectFile.DefaultHashSize);
+        if (!aCodeHashes.SequenceEqual(bCodeHashes))
+            return false;
+
+        return true;
+    }
+
+    public static CodeSignature Create(
+        (EmbeddedSignatureHeader Header, long FileOffset) embeddedSignature,
+        (CodeDirectoryHeader Header, long FileOffset) codeDirectory,
+        (byte[] identifier, long FileOffset) identifierPtr,
+        (byte[], long FileOffset) cdHashes,
+        (RequirementsBlob Header, long FileOffset) requirementsBlob,
+        (CmsWrapperBlob Header, long FileOffset) cmsWrapperBlob)
+    {
+        var cs = new CodeSignature();
+        cs._embeddedSignature = embeddedSignature;
+        cs._codeDirectory = codeDirectory;
+        cs._identifierPtr = identifierPtr;
+        cs._cdHashes = cdHashes;
+        cs._requirementsBlob = requirementsBlob;
+        cs._cmsWrapperBlob = cmsWrapperBlob;
+        cs._unrecognizedFormat = false;
+        return cs;
+    }
+
+    public static CodeSignature Read(MemoryMappedViewAccessor file, long fileOffset)
```

Comment:
Do we need this functionality? That is, reading out of a blob? For testing it seems like we could just validate that the bit pattern is correct. We shouldn't necessarily need to rehydrate the whole blob.

Reply:
We can't just do a byte-for-byte comparison since the hashes for the first page for two will be different if that's what you mean. We might be able to move this method to the test assembly, though.